docs/decisions/0040-OTel-semantic-convention.md

```diff
@@ -0,0 +1,460 @@
+- (Recommended)`gen_ai.response.finish_reasons`
+- (Recommended)`gen_ai.response.prompt_tokens`
+- (Recommended)`gen_ai.response.completion_tokens`
+
+The following events will be optionally attached to an activity:
+| Event name| Attribute(s)|
+|---|---|
+|`gen_ai.content.prompt`|`gen_ai.prompt`|
+|`gen_ai.content.completion`|`gen_ai.completion`|
+
+> The kernel must provide configuration options to disable these events because they may contain PII.
+> See the [Semantic conventions for generative AI](https://github.com/open-telemetry/semantic-conventions/tree/main/docs/gen-ai) for requirement level for these attributes.
+
+## Where do we create the activities
+
+It is crucial to establish a clear line of responsibilities, particularly since certain service providers, such as the Azure OpenAI SDK, have pre-existing instrumentation (in which case SK won't need to track LLM requests). Our objective is to position our activities as close to the model level as possible to promote a more cohesive and consistent developer experience.
+
+```mermaid
+block-beta
+columns 1
+    Models
+    blockArrowId1<["&nbsp;&nbsp;&nbsp;"]>(y)
+    block:Clients
+        columns 3
+        ConnectorTypeClientA["Instrumented client SDK<br>(i.e. Azure OpenAI client)"]
+        ConnectorTypeClientB["Un-instrumented Client SDK"]
+        ConnectorTypeClientC["Custom client on REST API<br>(i.e. HuggingFaceClient)"]
+    end
+    Connectors["Connectors/AI Services"]
+    blockArrowId2<["&nbsp;&nbsp;&nbsp;"]>(y)
+    SemanticKernel["Semantic Kernel"]
+    block:Kernel
+        Function
+        Planner
+        Agent
+    end
+```
+
+> Semantic Kernel also supports other types of connectors for memories/vector databases. We will discuss instrumentations for those connectors in a separate ADR.
+
+> Note that this will not change our approaches to [instrumentation for planners](./0025-planner-telemetry-enhancement.md). We may modify or remove some of the meters we created previously, which will introduce breaking changes.
+
+In order to keep the activities as close to the model level as possible, we should keep them at or below the connector level.
+
+### Out of scope
+
+These services will be discuss in the future:
+
+- Memory/vector database services
+- Audio to text services (`IAudioToTextService`)
+- Embedding services (`IEmbeddingGenerationService`)
+- Image to text services (`IImageToTextService`)
+- Text to audio services (`ITextToAudioService`)
+- Text to image services (`ITextToImageService`)
+
+## Considered Options
+
+- Instrumentation implementations
+  - Static class
+  - Dependency injection
+- Where do we create the activities
+  - Add instrumentations at the connector/service level
+  - Add instrumentations at the client level
+
+### Instrumentation implementations
+
+#### Static class `ModelDiagnostics`
+
+This class will live under `SemanticKernel.Abstractions/AI`.
+
+`ModelDiagnostics`
+
+```C#
+// Example
+public static class ModelDiagnostics
```

Comment:
Why does this need to be public? We'd have a lot more flexibility, including for how it evolves, if it's just an implementation detail built into whatever needs it.

Reply:
All connectors are individual projects, so they need to be able to reference this class.

Reply:
It needn't be public. It could be a shared file built into each project, for example.

Reply:
Sure, that is also an option. If we link this file in all connector projects, this file probably can't live under `SemanticKernel.Abstractions/AI`. `InternalUtilities` is a more suitable place.